### ISSUE TITLE
Guaranteed crash after about a minute of play

### ISSUE THREAD

**Issue by N1ckn1ght** [farmersrice/saltzero#3]
I'm trying to play using HumanVsRobotTest.py script provided. But after about a minute of play, it crashes no matter what. Even if no actions made at all.
Maybe you could help me with this issue? Or add a list of dependencies if I'm using some wrong version of the lib. I genuinely have no idea by now.
I couldn't launch the script on newest version of tensorflow at all, so I used these: `pip install tensorflow==2.10.0 numpy==1.23.5`

Windows 10
Python 3.10.11

Package                      Version
---------------------------- ----------
absl-py                      2.3.1
astunparse                   1.6.3
cachetools                   6.2.4
certifi                      2025.11.12
charset-normalizer           3.4.4
dotdict                      0.1
flatbuffers                  25.12.19
gast                         0.4.0
google-auth                  2.45.0
google-auth-oauthlib         0.4.6
google-pasta                 0.2.0
grpcio                       1.76.0
h5py                         3.15.1
idna                         3.11
keras                        2.10.0
Keras-Preprocessing          1.1.2
libclang                     18.1.1
Markdown                     3.10
markdown-it-py               4.0.0
MarkupSafe                   3.0.3
mdurl                        0.1.2
mem-top                      0.2.1
ml_dtypes                    0.5.4
namex                        0.1.0
numpy                        1.23.5
oauthlib                     3.3.1
opt_einsum                   3.4.0
optree                       0.18.0
packaging                    25.0
pillow                       12.0.0
pip                          25.3
protobuf                     3.19.6
pyasn1                       0.6.1
pyasn1_modules               0.4.2
Pygments                     2.19.2
requests                     2.32.5
requests-oauthlib            2.0.0
rich                         14.2.0
rsa                          4.9.1
setuptools                   80.9.0
six                          1.17.0
tensorboard                  2.10.1
tensorboard-data-server      0.6.1
tensorboard-plugin-wit       1.8.1
tensorflow                   2.10.0
tensorflow-estimator         2.10.0
tensorflow-io-gcs-filesystem 0.31.0
termcolor                    3.3.0
typing_extensions            4.15.0
urllib3                      2.6.2
Werkzeug                     3.1.4
wheel                        0.45.1
wrapt                        2.0.1

**Comment by farmersrice** [comment]
Can you provide more details on what the crash looks like? Is there a stack trace? What gpu do you have? I no longer have the computer which this previously ran on, so setting it up again could take some time.

> I couldn't launch the script on newest version of tensorflow at all

The tensorflow API is constantly changing between versions and this is not very good. Back when I built this 7y ago I didn't know to include a requirements file with versions, oops. We learn every day.

My understanding is Windows 10 is no longer receiving any updates. Probably unrelated, but who knows.

**Comment by farmersrice** [comment]
I just set it up successfully and lost a game after around 6 minutes of play. No crash on my end. I used `pip install tensorflow==2.12.0` on python 3.11.9, windows 11. 

For some reason it's not using the gpu though, I'm pretty sure it used to do that automatically.

The stack trace would help a lot. If you do get it working, also try the v0.0.3, because someone told me that one is actually stronger.

**Comment by N1ckn1ght** [comment]
Sorry for not providing stack trace. Said crash was killing python process, leaving the cmd open.

I just downloaded SaltZero again and set up python 3.11.9 venv with `tensorflow==2.12.0` (also dotdict, mem-top) and it **worked flawlessly**. Thank you! **This probably should be in README**.

_Still not on GPU, yes. I wonder if this could be related to CUDA version. For the record, works quick on CPU, but I'm not sure whether you had to cut engine strength for that, or not._

>also try the v0.0.3, because someone told me that one is actually stronger.

I wanted to test my hand-crafted evaluation against the real deal, and sz v0.0.4 annihilates it just fine. But I'll keep in mind to check against v0.0.3 ver. if I'll find ways to upgrade it =)

**Comment by farmersrice** [comment]
Nice. I looked back over the code and it really was so experimental and unpolished. Unfortunate that it was hard to run for you. If I come back to this project again, I would likely rewrite the whole thing from scratch in pytorch. I worked with pytorch a few years after this project and found that the docs were so much cleaner, they didn't keep changing the api unlike tensorflow, they had good C++ support (TF completely undocumented), etc. 

Not using the GPU in the `HumanVsRobotTest.py` is ok, because it is a fixed amount of compute at 400 playouts. It does not reduce the strength. 

I saw you had an engine on your github profile, you should let me know how it does in a 1v1.

**Comment by N1ckn1ght** [comment]
Everything's good, you helped me run it!

>I saw you had an engine on your github profile, you should let me know how it does in a 1v1.

Yeah, as I said, Vale (my engine) v0.1.1 loses to SaltZero v0.0.4, it gets annihilated even when playing as X. Vale is kinda naive because it has handcrafted evaluation, but I'm thinking of fine-tuning it.
I uploaded Vale to Codingame arena and got #414 placement, kinda falls behind the approximate #2 placement you have in your README =) Still, the coding challenge is fun.
I am actually amazed you replied in the repo you created 5 years ago.

**Comment by farmersrice** [comment]
The thing about that is NinjaDoggy's bot was originally 2nd place there when I asked him for the source and ran the test. But now it's 127th place. I'm not sure how it stacks up to the current generation of competitors.
